{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "from matplotlib.colors import LogNorm\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "import scipy\n",
    "\n",
    "import gseapy as gp\n",
    "lib_names = gp.get_library_name(organism='Human')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Module-level analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save gsea terms dict for all celltypes\n",
    "path = 'data'\n",
    "file = f'{path}/gsea-terms_all-celltypes_dict.pkl'\n",
    "\n",
    "# load gsea terms dict for all celltypes\n",
    "with open(file, 'rb') as f:\n",
    "   multicelltype_gsea_term_sets = pickle.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Keyword search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dataframe\n",
    "cols = []\n",
    "for k,v in multicelltype_gsea_term_sets.items():\n",
    "    comps = np.sort(v['component'].unique())\n",
    "    cols.extend([k+ ' ' + str(c+1) for c in comps])\n",
    "\n",
    "corr_threshold = 0.1\n",
    "max_superset_count = 0\n",
    "\n",
    "def get_kw_names(kw_df, kws):\n",
    "    if(type(kws) is not list):\n",
    "        return\n",
    "\n",
    "    def locate_kw(row, kw):\n",
    "        return kw in row.name.lower()\n",
    "\n",
    "    for kw in kws:\n",
    "        for celltype in multicelltype_gsea_term_sets.keys():\n",
    "            celltype_gsea_terms = multicelltype_gsea_term_sets[celltype]\n",
    "\n",
    "            celltype_gsea_terms = celltype_gsea_terms[np.abs(celltype_gsea_terms['pseudodx_corr'])>corr_threshold]\n",
    "            celltype_gsea_terms = celltype_gsea_terms[celltype_gsea_terms['superset_count']<=max_superset_count]\n",
    "\n",
    "            kw_mask = celltype_gsea_terms.apply(locate_kw, args=(kw,), axis=1)\n",
    "\n",
    "            gene_set_names = pd.Series([t[:-1] for t in celltype_gsea_terms.index[kw_mask]], dtype='object')\n",
    "            gene_set_comps = np.array([celltype + ' ' + str(c+1) for c in celltype_gsea_terms['component'][kw_mask]])\n",
    "            # color by nes (enrichment score) or fdr\n",
    "            gene_set_nes = -np.log(celltype_gsea_terms['fdr'][kw_mask].values + 1e-6) * np.sign(celltype_gsea_terms['nes'][kw_mask].values)\n",
    "\n",
    "            # add gene sets\n",
    "            for gs in gene_set_names.unique():\n",
    "                if(gs not in kw_df.index):\n",
    "                    kw_df = pd.concat((kw_df, pd.DataFrame(data=0, columns=kw_df.columns,index=pd.Index([gs]), dtype=int)), axis=0)\n",
    "                \n",
    "                for col in kw_df.columns:\n",
    "                    gs_mask = np.logical_and(gene_set_comps==col, gene_set_names.values==gs)\n",
    "                    kw_df.loc[gs, col] += np.sum(gene_set_nes[gs_mask])\n",
    "    \n",
    "    return kw_df\n",
    "\n",
    "kw_df = pd.DataFrame(columns=cols, dtype=int)\n",
    "\n",
    "# construct dataframe of keyword hits\n",
    "kw_df = get_kw_names(kw_df, ['microglia', 'mapk', 'inflamm', 'tumor necrosis', \\\n",
    "                            'mhc', 'toll-like', 'oligod', 'myelin', 'alzheimer', \\\n",
    "                            'amyloid', 'lipid', 'cholesterol', 'neuron', 'actin', \\\n",
    "                            'apopto', 'phagocyt', 'copper'])\n",
    "\n",
    "# remove gene set codes (GO, Wikipathways, Panther)\n",
    "kw_df.index = kw_df.index.to_series().str.split(' WP', expand=True)[0].str.split(\" \\(GO\", expand=True)[0].str.split(\" P0\", expand=True)[0].str.title().values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot results heatmap\n",
    "ccmap = 'coolwarm'\n",
    "\n",
    "kw_df = kw_df.replace(0, np.nan)\n",
    "\n",
    "plt.figure(figsize=(4,25))\n",
    "axr = sns.heatmap(kw_df, cmap=ccmap, linewidths=1, linecolor='k', square=True, cbar=False, vmin=-5, vmax=5 )\n",
    "axr.tick_params(left=False, right=True, top=False, labelleft=False, labelright=True, labeltop=False, rotation=0)\n",
    "plt.xticks(rotation=90);\n",
    "axr.tick_params(labelsize=12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GWAS search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2021 GWAS (38 loci)\n",
    "# gene list from: Wightman, D. P. et al. A genome-wide association study \n",
    "# with 1,126,563 individuals identifies new risk loci for Alzheimer’s disease. Nat. Genet. 53, 1276–1282 (2021).\n",
    "# renamed genes (alternative): INPPD5, EPHA1-AS1 (EPHA1), TSPOAP1-AS1 (SUPT4H1 (or RNF43 or BZRAP1)), \n",
    "gwas_goi = ['AGRN','CR1','NCK2','BIN1','INPPD5','CLNK','TNIP1','HAVCR2','HLA-DRB1','TREM2','CD2AP',\n",
    "    'TMEM106B','ZCWPW1', 'NYAP1','EPHA1','CLU','SHARPIN','USP6NL', 'ECHDC3','CCDC6','MADD', 'SPI1','MS4A4A','PICALM',\n",
    "    'SORL1','FERMT2','RIN3','ADAM10','APH1B','SCIMP', 'RABEP1','GRN','ABI3','SUPT4H1','ACE','ABCA7','APOE','NTN5',\n",
    "    'CD33','LILRB2','CASS4','APP',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rank pathways by number of gwas hits\n",
    "def count_pathway_gwas(row):\n",
    "    return len(row['AD GWAS genes'])\n",
    "\n",
    "def test_pathway_gwas(list, gene):\n",
    "    if(len(list)==0):\n",
    "        return False\n",
    "    return gene in list\n",
    "\n",
    "fdr_threshold = 0.05\n",
    "corr_threshold = 0.1\n",
    "max_superset_count = 0 \n",
    "\n",
    "gwas_celltype_count_dict = {}\n",
    "gwas_celltype_dict = {hit:[] for hit in gwas_goi}\n",
    "\n",
    "for celltype, gsea_terms_celltype in multicelltype_gsea_term_sets.items():\n",
    "    gsea_terms_celltype = gsea_terms_celltype[gsea_terms_celltype['fdr']<fdr_threshold]\n",
    "    gsea_terms_celltype = gsea_terms_celltype[gsea_terms_celltype['pseudodx_corr']>corr_threshold]\n",
    "    gsea_terms_celltype = gsea_terms_celltype[gsea_terms_celltype['superset_count']<=max_superset_count]\n",
    "    for comp in range(gsea_terms_celltype['component'].unique().size):\n",
    "        # gsea_terms = gsea_terms[~gsea_terms.index.duplicated()]\n",
    "        gsea_terms_celltype_comp = gsea_terms_celltype[gsea_terms_celltype['component']==comp].copy()\n",
    "        gsea_terms_celltype_comp['gwas_count'] = gsea_terms_celltype_comp.apply(count_pathway_gwas, axis=1)\n",
    "\n",
    "        # look at which gwas genes are present in top pathways across celltypes\n",
    "        gwas_celltype_count_dict[f'{celltype} {comp+1}'] = []\n",
    "        for hit in gwas_goi:\n",
    "            gwas_celltype_count_dict[f'{celltype} {comp+1}'].append(gsea_terms_celltype_comp['ledge_genes'].apply(test_pathway_gwas, args=[hit]).sum())\n",
    "            gwas_celltype_dict[hit].extend([f'{celltype} comp-{comp+1}: {x}' for x in gsea_terms_celltype_comp.index[gsea_terms_celltype_comp['ledge_genes'].apply(test_pathway_gwas, args=[hit])].values.tolist()])\n",
    "\n",
    "gwas_celltype_df = pd.DataFrame(gwas_celltype_count_dict, index=gwas_goi)\n",
    "\n",
    "df_plot = gwas_celltype_df[gwas_celltype_df.sum(axis=1)>0] # drop genes with no hits\n",
    "df_plot = df_plot.loc[(df_plot>0).sum(axis=1).sort_values(ascending=False).index] # sort columns (genes) by number number of component hits\n",
    "\n",
    "plt.figure(figsize=(20,8))\n",
    "sns.heatmap(df_plot.T, cmap='Reds', annot=True, norm=LogNorm(), cbar=False, linewidths=1, linecolor='k')\n",
    "plt.xticks(rotation=60, ha='center')\n",
    "plt.title('Number of pathways containing top Alzheimer GWAS hits')\n",
    "plt.xlabel('GWAS hits')\n",
    "plt.ylabel('Celltype');"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Celltype-celltype interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fdr_threshold = 0.05\n",
    "corr_threshold = 0.1\n",
    "\n",
    "# look at coordination between top N pathways (by pseudodx correlation) between celltypes\n",
    "def cross_correlate_top_pathways(gsea_terms, cell1, comp1, cell2, comp2, rank_column, metric_list_column, n_compare=10):\n",
    "    gsea_terms_celltype1 = gsea_terms[cell1].sort_values(by=rank_column, ascending=True)\n",
    "    gsea_terms_celltype1 = gsea_terms_celltype1[gsea_terms_celltype1['component']==comp1]\n",
    "    gsea_terms_celltype1 = gsea_terms_celltype1[gsea_terms_celltype1['fdr']<fdr_threshold]\n",
    "    gsea_terms_celltype1 = gsea_terms_celltype1[gsea_terms_celltype1['pseudodx_corr']>corr_threshold]\n",
    "    \n",
    "    gsea_terms_celltype2 = gsea_terms[cell2].sort_values(by=rank_column, ascending=True)\n",
    "    gsea_terms_celltype2 = gsea_terms_celltype2[gsea_terms_celltype2['component']==comp2]\n",
    "    gsea_terms_celltype2 = gsea_terms_celltype2[gsea_terms_celltype2['fdr']<fdr_threshold]\n",
    "    gsea_terms_celltype2 = gsea_terms_celltype2[gsea_terms_celltype2['pseudodx_corr']>corr_threshold]\n",
    "    \n",
    "    n_compare = min(n_compare, min(gsea_terms_celltype1.shape[0], gsea_terms_celltype2.shape[0]))\n",
    "\n",
    "    r_list = []\n",
    "    for i1 in range(n_compare):\n",
    "        for i2 in range(n_compare):\n",
    "            pathway1_by_id = np.array(gsea_terms_celltype1.iloc[i1][metric_list_column])\n",
    "            pathway2_by_id = np.array(gsea_terms_celltype2.iloc[i2][metric_list_column])\n",
    "            \n",
    "            # pearson\n",
    "            nas = np.logical_or(np.isnan(pathway1_by_id), np.isnan(pathway2_by_id))\n",
    "            r = scipy.stats.pearsonr(pathway1_by_id[~nas], pathway2_by_id[~nas])[0]\n",
    "            \n",
    "            r_list.append((f'{cell1} {comp1+1}',f'{cell2} {comp2+1}',gsea_terms_celltype1.index[i1],gsea_terms_celltype2.index[i2],r))\n",
    "    \n",
    "    return r_list\n",
    "\n",
    "gsea_celltypes = list(multicelltype_gsea_term_sets.keys())\n",
    "celltype_n_comps = {cell:multicelltype_gsea_term_sets[cell]['component'].unique().shape[0] for cell in gsea_celltypes}\n",
    "gsea_celltypes_comps = [f'{cell} {comp+1}' for cell in gsea_celltypes for comp in range(celltype_n_comps[cell])]\n",
    "celltype_pair_df = pd.DataFrame(0, index=gsea_celltypes_comps, columns=gsea_celltypes_comps)\n",
    "\n",
    "pathway_pair_columns = ['celltype1', 'celltype2', 'pathway1', 'pathway2', 'correlation']\n",
    "pathway_pair_df = pd.DataFrame(columns=pathway_pair_columns)\n",
    "\n",
    "n_celltypes = len(gsea_celltypes)\n",
    "for i_cell1 in range(n_celltypes):\n",
    "    for i_cell2 in range(i_cell1+1,n_celltypes):\n",
    "        cell1 = gsea_celltypes[i_cell1]\n",
    "        cell2 = gsea_celltypes[i_cell2]\n",
    "\n",
    "        for comp1 in range(celltype_n_comps[cell1]):\n",
    "            for comp2 in range(celltype_n_comps[cell2]):\n",
    "                # evaluate correlation between pairs of top pathways\n",
    "                r_list = cross_correlate_top_pathways(multicelltype_gsea_term_sets, cell1, comp1, cell2, comp2, \n",
    "                                                      rank_column='fdr', metric_list_column='dx_pred_score', n_compare=50)\n",
    "                \n",
    "                pathway_pair_df_tmp = pd.DataFrame(r_list, columns=pathway_pair_columns)\n",
    "\n",
    "                # save median absolute correlation for celltype pair\n",
    "                celltype_pair_df.loc[f'{cell1} {comp1+1}', f'{cell2} {comp2+1}'] = np.median(np.abs(pathway_pair_df_tmp['correlation']))\n",
    "\n",
    "                # save pathway pairs\n",
    "                pathway_pair_df = pd.concat([pathway_pair_df, pathway_pair_df_tmp], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mask out upper triangle of dataframe\n",
    "celltype_pair_df = celltype_pair_df.replace(0,np.nan)\n",
    "\n",
    "# set color of None values\n",
    "cmap = cm.get_cmap('Reds').copy()\n",
    "cmap.set_bad('w')\n",
    "\n",
    "plt.figure(figsize=(10,10))\n",
    "plt.imshow(celltype_pair_df.T, cmap=cmap)\n",
    "plt.title('Median absolute correlation between top pathway pairs', fontsize=16)\n",
    "plt.xticks(ticks=np.arange(len(gsea_celltypes_comps)), labels=gsea_celltypes_comps, rotation=90, fontsize=14)\n",
    "plt.yticks(ticks=np.arange(len(gsea_celltypes_comps)), labels=gsea_celltypes_comps, fontsize=14)\n",
    "plt.colorbar()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
